### PR TITLE
Course performace anlysis'

### PR DESCRIPTION
**Pull request by MtheDV**

### CODE REVIEW

**Comment by MtheDV** [comment]
@nononovia lgty?

File: scripts/course_performace_analysis/course_performace_analysis.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
+import json
```
Comment (carson-ricca) [inline scripts/course_performace_analysis/course_performace_analysis.py:1]: Main question, is there going to be another file in the root directory to run this script like the way we have done it for our other scripts?

File: scripts/course_performace_analysis/course_performace_analysis.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+import matplotlib.pyplot as plt
+import pandas as pd
+import scipy.stats
+import seaborn as sns
+import scipy as sci
+
+
+def course_performance_analysis(GRADEBOOK_PATH, QUIZSCOREJSON_PATH):
+    sns.color_palette('bright')
+    file = pd.read_csv(GRADEBOOK_PATH)
```
Comment (carson-ricca) [inline scripts/course_performace_analysis/course_performace_analysis.py:14]: If we do have a file in root that runs this then the `read_csv()` should be brought into that file instead, similar to what was done with the clustering stuff.

File: scripts/course_performace_analysis/course_performace_analysis.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    data448_ids = file['ID']
```
Comment (carson-ricca) [inline scripts/course_performace_analysis/course_performace_analysis.py:15]: All of these IDs could be put in a schema instead, so we can avoid the string literals here.

File: scripts/course_performace_analysis/course_performace_analysis.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    quiz_ids = []
+    final_score = file['Overall Final Score']
+
+    overall_pre_test = file['Overall Pre-Tests (880658)']
+    overall_post_test = file['Overall Post-Tests (880660)']
+    overall_pre_post_grade = 40
```
Comment (carson-ricca) [inline scripts/course_performace_analysis/course_performace_analysis.py:21]: where does value for this constant come from?

File: scripts/course_performace_analysis/course_performace_analysis.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+    # Actual Score mean and std
+    current_mean = final_score.mean()
+    current_standard_dev = final_score.std()
+
+    w_2016_avg, w_2016_std = 79.93, 7.9
+    w_2017_avg, w_2017_std = 78.70, 16.2
+    w_2018_avg, w_2018_std = 85.97, 8.8
+    w_2019_avg, w_2019_std = 83.81, 10.0
+    w_2020_retrieve_avg, w_2020_retrieve_std = 90.80, 10.5
```
Comment (carson-ricca) [inline scripts/course_performace_analysis/course_performace_analysis.py:105]: Assuming these values had to be hardcoded?

**Review by carson-ricca — CHANGES_REQUESTED** [review]

**Comment by carson-ricca** [comment]
Also conflicts should be resolved.

File: scripts/course_performace_analysis/course_performace_analysis.py (hunk 5 above)
Comment (MtheDV) [reply scripts/course_performace_analysis/course_performace_analysis.py:105]: yes, they were taken from the UBC website that provided previous years grades

File: scripts/course_performace_analysis/course_performace_analysis.py (hunk 4 above)
Comment (MtheDV) [reply scripts/course_performace_analysis/course_performace_analysis.py:21]: the pre and post tests are both worth 20% (gradebook) of the grade, therefore we need to work with the combined total

File: scripts/course_performace_analysis/course_performace_analysis.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    first_attempt_legend = pch.Patch(color=data_graph_colors[1], label='First Attempt Score')
+    ax1.legend(handles=[score_legend, first_attempt_legend])
+    fig.tight_layout()
+    plt.savefig(f'data/graphs/overall_vs_first_attempt.png')
+    plt.close()
+
+    # 2) Histogram comparison plot for Actual score and First attempt only score
+    fig, (ax1) = plt.subplots(1)
+    sns.histplot(final_score, kde=True, bins=30, line_kws={'linewidth': 1}, ax=ax1).set(
+        title='Overall Score vs First Attempt Score',
```
Comment (nononovia) [inline scripts/course_performace_analysis/course_performace_analysis.py:151]: How is the first attempt reflected in this plot? To me this plot just seems like a histogram for overall course performance. The title seems a bit misleading?

File: scripts/course_performace_analysis/course_performace_analysis.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    df_student_grade_first_attempt = pd.DataFrame(
+        columns=['DATA448_ID', 'total_score', 'total_possible_points', 'final_score', 'first_attempt_final_score']
+    )
+    df_average_student_attempts = pd.DataFrame(
+        columns=['DATA448_ID', 'QUIZ_ID', 'attempt', 'final_score']
+    )
+
+    # Get first attempt only quiz mark out of JSON files
+    number_of_quizzes = 0
+    for i in os.listdir(QUIZSCOREJSON_PATH):
```
Comment (nononovia) [inline scripts/course_performace_analysis/course_performace_analysis.py:36]: What is your file structure like? Maybe create empty __init__.py files in the directories you create so your file structure is included in the repo (?

File: schemas/gradebook.py [hunk 8]
```diff
@@ -9,3 +10,7 @@ class GradeBookSchema:
     OVERALL_PRE_POST_TESTS_SCORE = 'Pre/Post Tests Final Score'
     OVERALL_READING_LOGS_SCORE = 'Reading Logs Final Score'
     OVERALL_COURSE_SCORE = 'Final Score'
+    FINAL_SCORE = 'Overall Final Score'
+    PRE_TEST_SCORE = 'Overall Pre-Tests (880658)'
+    POST_TEST_SCORE = 'Overall Post-Tests (880660)'
+    COMBINED_PRE_POST_WORTH = 40
```
Comment (Opeyem1a) [inline schemas/gradebook.py:16]: This feels like a constant to be stored elsewhere and not a schema thing

File: scripts/course_performace_analysis/course_performace_analysis.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+import pandas as pd
+import scipy.stats
+import seaborn as sns
+
+from schemas import GradeBookSchema
+
+
+def course_performance_analysis(gradebook, QUIZSCOREJSON_PATH):
+    sns.color_palette('bright')
+    file = gradebook
```
Comment (Opeyem1a) [inline scripts/course_performace_analysis/course_performace_analysis.py:15]: I think keeping it named `gradebook` going forward makes the code more readable

File: scripts/course_performace_analysis/course_performace_analysis.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    )
+    df_average_student_attempts = pd.DataFrame(
+        columns=['DATA448_ID', 'QUIZ_ID', 'attempt', 'final_score']
+    )
+
+    # Get first attempt only quiz mark out of JSON files
+    number_of_quizzes = 0
+    for i in os.listdir(QUIZSCOREJSON_PATH):
+        if i.endswith('.json'):
+            number_of_quizzes += 1
```
Comment (Opeyem1a) [inline scripts/course_performace_analysis/course_performace_analysis.py:38]: I didn't review further than this, just leaving it as a marker

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

File: schemas/gradebook.py (hunk 8 above)
Comment (carson-ricca) [reply schemas/gradebook.py:16]: Yeah this value might work better as a constant...

File: scripts/course_performace_analysis/course_performace_analysis.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+            full_path = os.path.join(QUIZSCOREJSON_PATH, i)
+            with open(full_path, 'r') as f:
+                file = f.read()
+                json_file = json.loads(file)
+                if json_file:
+                    if json_file[0]['quiz_points_possible'] > 0:
```
Comment (carson-ricca) [inline scripts/course_performace_analysis/course_performace_analysis.py:44]: Wondering if all the ids such as `quiz_points_possible` should be in a schema or set as constants cause they seemed to be used a lot throughout the following code.

**Review by carson-ricca — CHANGES_REQUESTED** [review]

File: schemas/gradebook.py (hunk 8 above)
Comment (nononovia) [reply schemas/gradebook.py:16]: Why do these work better as constants? They are column names just like in other schemas? (I'm probably just confused about what schemas actually do, but seems like in schema/attendance.py only constants are stored as well..?)

Comment (carson-ricca) [reply schemas/gradebook.py:16]: I think for me it was just the `COMBINED_PRE_POST_WORTH` it didn't seem like a column name, but if it's a column name then it should be chill.

**Review by carson-ricca — DISMISSED** [review]

**Review by carson-ricca — DISMISSED** [review]
Looks good to me

**Review by nononovia — APPROVED** [review]

**Review by Opeyem1a — APPROVED** [review]

**Pull request merged**